kafka/consumer.go

```diff
@@ -153,39 +174,80 @@ func NewConsumer(cfg ConsumerConfig) (*Consumer, error) {
 		cfg.MaxPollRecords = 100
 	}
 	return &Consumer{
-		cfg:      cfg,
-		client:   client,
-		consumer: consumer,
-		tracer:   cfg.tracerProvider().Tracer("kafka"),
+		cfg:        cfg,
+		client:     client,
+		consumer:   consumer,
+		closed:     make(chan struct{}),
+		running:    make(chan struct{}),
+		forceClose: forceClose,
+		stopPoll:   func() {},
+		tracer:     cfg.tracerProvider().Tracer("kafka"),
 	}, nil
 }
 
 // Close the consumer, blocking until all partition consumers are stopped.
 func (c *Consumer) Close() error {
-	if !c.closed.CompareAndSwap(false, true) {
-		return nil
-	}
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.client.Close()
-	c.consumer.wg.Wait() // Wait for all the goroutines to exit.
+	select {
+	case <-c.closed:
+	default:
+		close(c.closed)
+		// Cancel the context used in client.PollRecords, triggering graceful
+		// cancellation.
+		c.stopPoll()
+		stopped := make(chan struct{})
+		go func() {
+			defer close(stopped)
+			// Close all partition consumers first to ensure there aren't any
+			// records being processed while the kgo.Client is being closed.
+			// Also ensures that commits can be issued after the records are
+			// processed when AtLeastOnceDelivery is configured.
+			c.consumer.close()
+		}()
+		// Wait for the consumers to process any in-flight records, or cancel
+		// the underlying processing context if they aren't stopped in time.
+		select {
+		case <-time.After(c.cfg.ShutdownGracePeriod): // Timeout
+			c.forceClose(fmt.Errorf(
+				"consumer: close: timeout waiting for consumers to stop (%s)",
+				c.cfg.ShutdownGracePeriod.String(),
+			))
+		case <-stopped: // Stopped within c.cfg.ShutdownGracePeriod.
+		}
+		c.client.Close() // Last, close the `kgo.Client`
+	}
+	c.consumer.wg.Wait() // Wait for consumer goroutines to exit.
 	return nil
 }
 
 // Run the consumer until a non recoverable error is found:
 //   - ErrCommitFailed.
 //
-// To shut down the consumer, cancel the context, or call consumer.Close().
+// To shut down the consumer, call consumer.Close() or cancel the context.
+// Calling `consumer.Close` is advisable to ensure graceful shutdown and
+// avoid any records from being lost (AMOD), or processed twice (ALOD).
+//
 // If called more than once, returns `apmqueue.ErrConsumerAlreadyRunning`.
 func (c *Consumer) Run(ctx context.Context) error {
-	if !c.running.CompareAndSwap(false, true) {
+	c.mu.Lock()
+	select {
+	case <-c.running:
+		c.mu.Unlock()
 		return apmqueue.ErrConsumerAlreadyRunning
+	default:
+		close(c.running)
 	}
+	// Create a new context from the passed context, used exclusively for
+	// kgo.Client.* calls. c.stopFetch is called by consumer.Close() to
+	// cancel this context as part of the graceful shutdown sequence.
+	var clientCtx context.Context
+	clientCtx, c.stopPoll = context.WithCancel(ctx)
```

Comment:
Because `stopPoll` is only set after `Run` is called, calling `Stop` before `Run` means that the poll context won't be closed. Should we instead start a goroutine that waits for `c.closed` and cancels the context?

Reply:
sorry, but I'm afraid I don't follow. The `clientCtx` is only used if `Run()` is called, otherwise it's not used. `c.stopPoll` is set to an empty function and on close would be a no-op.

Reply:
```go
consumer.Close()
consumer.Run() <- clientCtx is not cancelled because Run was called after Close returned
```

I suppose it doesn't matter though, because Close will close the kgo.Client, causing fetch to return context.Canceled, breaking the loop.